init/src/cli.rs

```diff
@@ -0,0 +1,377 @@
+                            match value {
+                                Some(env_value) => {
+                                    let (key_env, value_env) = to_env_pair(env_value)?;
+                                    args.override_env
+                                        .insert(key_env.to_os_string(), value_env.to_os_string());
+                                }
+                                None => return Err("No environment variables passed".into()),
+                            }
+                        }
+                        Ok('w') => {
+                            let value = shorts.next_value_os();
+                            match value {
+                                Some(dir) => {
+                                    args.workdir = Some(PathBuf::from(dir));
+                                }
+                                None => return Err("No working directory passed".into()),
+                            }
+                        }
+                        Ok('t') => {
+                            let timeout_arg = shorts.next_value_os();
+                            match timeout_arg {
+                                Some(timeout_arg) => {
+                                    let timeout_opt = timeout_arg.to_str();
+                                    match timeout_opt {
+                                        Some(timeout) => {
+                                            let timeout = str::parse(timeout)?;
+                                            args.timeout = timeout;
+                                        }
+                                        None => {
+                                            return Err(format!(
+                                                "Invalid timeout {timeout_arg:?} passed"
+                                            )
+                                            .into())
+                                        }
+                                    }
+                                }
+                                None => return Err("No timeout passed".into()),
+                            }
+                        }
+                        _ => return Err(format!("Unexpected flag: -{short:?}").into()),
+                    }
+                }
+            } else {
+                seen_child_process = true;
+                let value = arg
+                    .to_value_os()
+                    .to_owned()
+                    .into_string()
+                    .map_err(|o| format!("Invalid argument: {o:?}"))?;
+                args.child_process = Some(value.clone());
+                args.child_args.push(value);
+            }
+        } else {
+            let value = arg
+                .to_value_os()
+                .to_owned()
+                .into_string()
+                .map_err(|o| format!("Invalid argument: {o:?}"))?;
+            args.child_args.push(value)
+        }
+    }
+
+    Ok(args)
+}
```

Comment:
Wow, that's a lot more work than just normal `clap` :). I wouldn't feel bad about making `clap` a dependency instead if you'd prefer.

Reply:
> Wow, that's a lot more work than just normal clap :). I wouldn't feel bad about making clap a dependency instead if you'd prefer.

Yeah, it was frustating to use this after being used to `clap`. But our static binary size currently is much better than other projects like `tini` , so I still think it's worth it since `clap_lex` has zero dependencies.

Reply:
Got it. Not saying you should do a bunch of extra work, but I'd be curious if using clap directly results in much larger executables. I would hope most of the complexity in clap itself comes from the macros, which is a compile-time-only concept. But I could be completely mistaken.

Reply:
Thanks, I wll merge this PR as is. I will also create a prototype using clap and see how much difference in binary size it is. For reference this is the current binary size:

```
❯ ls -lah target/x86_64-unknown-linux-musl/release/pid1
.rwxr-xr-x sibi users 494 KB Sun Oct  1 12:10:41 2023  target/x86_64-unknown-linux-musl/release/pid1
```

Reply:
@snoyberg I made a prototype using `clap` and it does make the binary size to `658` KB. But this still comes lesser than `tini` init (which is around 850 KB) but not less than the Haskell's pid1 (where we anyway cheat using `upx` without which it goes to MB).

I think that ~164 KB increase in size is probably worth justifying given that it simplfies the code to a huge amount. WDYT ? If you are fine, I will open a new PR with clap changes.

Reply:
I agree that it’s worth it. Also interested what the difference is after using upx. Also, just to confirm, are these numbers after stripping the binary?

Reply:
> Also interested what the difference is after using upx.

The Haskell `pid1` binary gets transformed from 1.5 MB to 533 KB

The rust `pid1` binary gets transformed from ~680 KB to ~314 KB.

But overall, I think it's best to not use `upx`, because there has been scenario's where the Haskell binary did segfault after using it. Also, `upx` does compression - which it also has to decompress it during runtime.

> are these numbers after stripping the binary

Yes. Cargo provides nice integration with `strip` - so all I need to is provide the following tweak in the release profile: https://github.com/fpco/pid1-rs/pull/8/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R12